Synonym Graph Support (LUCENE-6664)

mattweber opened the pull request:
Integrate @mikemccand's patch from [LUCENE-6664](https://issues.apache.org/jira/browse/LUCENE-6664) into elasticsearch and add support for handling a graph token stream in match/multi-match queries.  

This PR still needs some work, ie. lots more tests, fix some TODO's, etc.  I wanted to get some feedback and see if you would even consider merging this before I spend much more time on it.  I do have it as a plugin as well but it would be much nicer in core since the plugin needs to fork Match and MultiMatch queries.

clintongormley commented:
heya @mattweber - thanks for the PR!

@mikemccand could you have a look at this please?

mikemccand commented on core/src/main/java/org/apache/lucene/util/XQueryBuilder.java:
```diff
@@ -0,0 +1,496 @@
+                    positionCount += positionIncrement;
+                } else {
+                    hasSynonyms = true;
+                    // TODO: better way to detect a graph token stream?
```
I *think* you could just pull the `PositionLengthAttribute` and if it has a value greater than 1 on any token, it is a graph.

mattweber replied:
Awesome, I like that much better.  I will make this change and set the type of the synonym back to `SYNONYM`.

mattweber replied:
Ok this doesn't work because the standard synonym token filter also sets the position length attribute which will cause the query builder to attempt to build the automaton from a non-graph token stream.

Maybe I just introduce a new `GraphTokenStreamAttribute` attribute that defaults to `false` and is only set to `true` when using a graph-aware token filter and we have a synonym with a position length greater than 1?

mattweber replied:
I set a flag using `FlagsAttribute` when the `SynonymGraphFilter` sets a synonym.  I can then detect a graph by looking for this flag along with position length > 1.

mikemccand commented on core/src/main/java/org/apache/lucene/util/XRollingBuffer.java:
```diff
@@ -0,0 +1,151 @@
+ *
+ * Temporary until LUCENE-6664.
+ */
+public abstract class XRollingBuffer<T extends XRollingBuffer.Resettable> {
```
Hmm, why did we need to fork this class from Lucene?  Is it only for the added `getBufferSize` public API?  If so, I think we can make this change directly in Lucene.  I'll do that ...

mattweber commented:
@mikemccand I saw that you recently committed the new rewrite code to `TermAutomatonQuery`.  I decided not to go that route because I wanted to make sure I could control the different mode/operators across the different tokens in each path.  I think everything is essentially a phrase using `TermAutomatonQuery` and in this approach we use whatever the original query terms were configured to use.  The `GraphTokenStreamFiniteStrings` class uses all the `TokenStream` to `Automaton` code from `TermAutomatonQuery`, maybe it would be good to extract that out so we can reuse it.  Probably not important until we wanted to get it into Lucene.

I have only integrated this forked query parser into the `Match` queries which really don't do anything other than analysis.  I will have to see what it takes to get it into `QueryString` and `SimpleQueryString` queries.

Yes, `RollingBuffer` was only forked due to making `getBufferSize` public.  Looks like you already handled that.  Also, only reason we needed to fork `QueryBuilder` is because the `createFieldQuery` method was final.  Is that something we could change?

mattweber commented:
While I am thinking about it, I want to note that `minimum_should_match` is not currently being applied correctly when we detect a graph query.  I need to look into it.

mattweber commented:
Also, I had initially tried to use `BytesTermAttribute` to reuse the `BytesRef` we had for each term instead of converting back to a string and using `CharTermAttribute`.  I kept getting NPE when used with the `CachingTokenFilter` in the query builder.  Not sure if that is something worth investigating as well.

mikemccand commented:
> Also, only reason we needed to fork QueryBuilder is because the createFieldQuery method was final. Is that something we could change?

I'll open a Lucene issue to explore that; looks like it was originally (born) final here: https://issues.apache.org/jira/browse/LUCENE-5261

> I decided not to go that route because I wanted to make sure I could control the different mode/operators across the different tokens in each path. I think everything is essentially a phrase using TermAutomatonQuery and in this approach we use whatever the original query terms were configured to use.

Ahh OK ... can you give an example where you would not want the separate paths to be treated as a phrase query?  I'm just trying to understand the use case; maybe we can improve `TermAutomatonQuery` to work better for such cases.

>  The GraphTokenStreamFiniteStrings class uses all the TokenStream to Automaton code from TermAutomatonQuery, maybe it would be good to extract that out so we can reuse it. Probably not important until we wanted to get it into Lucene.

I like this idea, basically just a static method to turn a `TokenStream` into the equivalent set of `Token[]` ... but yes we can wait I think.  We also have `TokenStreamToAutomaton`, but that puts each byte of each token as a transition.

> I kept getting NPE when used with the CachingTokenFilter in the query builder. Not sure if that is something worth investigating as well.

Hmm do you have more details about this?  A traceback?  Likely something was making the hard assumption that a token would always have a `CharTermAttribute`.  `CachingTokenFilter` looks ok on quick inspection, and `QueryBuilder.createFieldQuery` seems to only pull the `TermToBytesRefAttribute` so I'm not sure where it's breaking.

mikemccand commented:
I opened https://issues.apache.org/jira/browse/LUCENE-7560 to explore removing `final` from `QueryBuilder.createFieldQuery`.

mattweber commented:
@mikemccand thanks for opening that issue!  I will get you more info on the `BytesTermAttribute` and `CachingTokenFilter` NPE I found.  I believe it was during a clone of the attribute where it tried to do deep copy of the `BytesRef` but the `BytesRef` itself was null for some reason.

> Ahh OK ... can you give an example where you would not want the separate paths to be treated as a phrase query? I'm just trying to understand the use case; maybe we can improve TermAutomatonQuery to work better for such cases.

For example, non-phrase match query of "say what the fudge" using the AND operator. I don't want this to be forced into a phrase, I just want all terms to match.  Since we have a synonym also "say wtf" can also match, again not as a phrase.  Ideally, only the expanded multi-token synonym itself would be treated as a phrase, not the original search terms.  So "say wtf" would end up looking like:

```
AND(say, wtf)
AND(say, PHRASE(what the fudge))
```

That AND could also be an OR, or a PHRASE depending on how the user configured the match query.  I figured it would get pretty hairy with all the different combinations so treating everything the same using the user's configured query settings would be best.  Does this make sense?

s1monw commented:
I love the stuff happening here! Yet, if we do this in ES and later move it to lucene we might be fighting with BWC big time, but don't get me wrong I think we should do it this way. To prevent these problems I think we need to make it experimental AND we need a way for people to opt into these features. I thought about this for a while and I think we need to add a setting to enable this which is disabled by default. To make this explicit we can keep these settings private and have a dedicated / generic `experimental_features` plugin that does nothing but registers these settings. That way users need to install the plugin AND set the setting to `enabled/disabled` for each feature.   This will be explicit and they will get failures if they use the feature and the feature got removed or moved from `experimental` to `supported` @clintongormley WDYT

mattweber commented:
@mikemccand Ignore my issue about the NPE, I must have been doing something else wrong at the time.  I just pushed a commit that uses `BytesTermAttribute` which avoids the unnecessary string conversion.  

@simonw You guys tell me what I need to do and I will work on it.  This feature is a big one for me and having this in core (even as an experimental plugin) vs. my custom plugin is a big win.

mattweber commented:
@simonw Also, I think the opt-in is the fact that the user uses the `synonym_graph` token filter at all.  If they don't use it, nothing changes.  Is that enough?

mattweber commented:
Just pushed another commit that will allow us to properly handle `minimum_should_match`.

s1monw commented:
> @simonw Also, I think the opt-in is the fact that the user uses the synonym_graph token filter at all. If they don't use it, nothing changes. Is that enough?

I am not sure - people get surprised quickly and BWC layers are difficult. I'd rather have another step to take here for users but don't worry about it here for now

mikemccand commented:
> I opened https://issues.apache.org/jira/browse/LUCENE-7560 to explore removing final from QueryBuilder.createFieldQuery.

OK this is now done (for Lucene 6.4.0).

mikemccand commented:
>  So "say wtf" would end up looking like:

Ahh thank you for that example @mattweber, that makes perfect sense, and the simplicity of simply enumerating all matching phrases out of the `TokenStream` graph is really nice.  We can later think about how we could make use of TAQ, only for the "graph parts", but keep the other query parser operators for the non-graph parts.

I like the idea of a separate, explicit `experimental_features` plugin, but I agree we should not let that hold up this issue (which is already complex enough).  For this issue we should make it clear in the docs that this is experimental, and when we later have this plugin, we can at that point apply it to this feature.

Can we rename `SausageGraphFilter` to maybe `SquashGraphFilter` or `FlattenGraphFilter` or something?  The docs need to make it clear that the index analyzer must use `SquashGraphFilter`.

mattweber commented:
@mikemccand I will do the rename and start on some documentation.  Now that I fixed the minimum should match bug yesterday I think all the major issues are resolved.  Do you guys see any major problems that might hold up getting this in?

mikemccand commented:
Thanks @mattweber ... I don't see any problems now.  One minor thing: can you add `TODO` to the two forked Lucene classes to un-fork them once we upgrade to Lucene 6.4.0?

mattweber commented:
@mikemccand Should we just remove `SausageGraphFilter` from this PR?   Really, if we were going to use `SynonymGraphFilter` and then `SausageGraphFilter` at index time we might as well just use the normal `SynonymFilter`.  We can document that `SynonymGraphFilter` is meant to be used as a search analyzer only.

mikemccand commented:
@mattweber I think that's a good approach.

Long term, I expect that `SynonymGraphFilter` will replace `SynonymFilter`, but short term (this change) let's minimize the risk and focus only on query time synonyms.

mattweber commented:
@mikemccand Ok, removed `SausageGraphFilter` and added the `TODO` for Lucene 6.4.0.  On that note, we need more changes to the upstream `QueryBuilder`... 

Can the following `private` methods be changed to `protected` so we can access them from our overridden `createFieldQuery` method?

- `analyzeTerm`
- `analyzeMultiPhrase`
- `analyzePhrase`
- `analyzeBoolean`
- `analyzeMultiBoolean`

mikemccand commented:
> Can the following private methods be changed to protected so we can access them from our overridden createFieldQuery method?

I'll try :)  I'll reopen the Lucene issue.

mattweber commented:
@mikemccand Thanks!  I have made my forked version a copy of what I think it will look like in Lucene 6.4.0 assuming those methods get changed from `private` to `protected`.  I also moved all the graph detection logic out into overridden methods of `MatchQueryBuilder` to make un-forking easier.  

I think we are ready for review, should I squash everything?

mikemccand commented:
> I think we are ready for review, should I squash everything?

++ thank you.

mattweber commented:
@mikemccand ready to go.  Thanks!

mikemccand commented on core/src/main/java/org/apache/lucene/search/GraphQuery.java:
```diff
@@ -0,0 +1,117 @@
+        BooleanQuery.Builder q = new BooleanQuery.Builder();
+        q.setDisableCoord(true);
+        for (Query clause : queries) {
+            if (clause != null) {
```
Hmm why would we allow a null clause to be sent to `GraphQuery`?

mattweber replied:
Just a safety check as it is possible as the initial array of queries is populated via `createFieldQuery` which potentially returns a null query.  I can remove this check and do it before creating the `GraphQuery`.

mikemccand commented on core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java:
```diff
@@ -158,11 +158,12 @@ public IndexAnalyzers build(IndexSettings indexSettings) throws IOException {
         Map<String, AnalysisModule.AnalysisProvider<TokenFilterFactory>> tokenFilters = new HashMap<>(this.tokenFilters);
         /*
-         * synonym is different than everything else since it needs access to the tokenizer factories for this index.
+         * synonym and synonym_graph is different than everything else since they need access to the tokenizer factories for the index.
```
is different -> are different

mikemccand commented on docs/reference/analysis/tokenfilters/synonym-graph-tokenfilter.asciidoc:
```diff
@@ -0,0 +1,146 @@
+
+The `synonym_graph` token filter allows to easily handle synonyms, 
+including multi-word synonyms correctly during the analysis process. 
+
```
Could you link here to the blog post the explains the complexities behind graph token streams?  http://blog.mikemccandless.com/2012/04/lucenes-tokenstreams-are-actually.html

mattweber commented:
@mikemccand thanks for reviewing.  I addressed everything but the null check, let me know what you want to do there and I will get it fixed and re-squash this for you.

mattweber commented:
@mikemccand Removed the null check inside of `GraphQuery` and do it as we create the actual query from each token stream.  Please let me know if there is anything else you would like me to do.  Can we backport this into `5.x` and hopefully `5.1`?

mikemccand commented:
Thank you @mattweber this LGTM now; maybe @jpountz could review from the ES query parsing standpoint?  I'm less familiar w/ that side of things.

I think 5.x makes sense; I'm unsure of the timing for 5.1.0, but I agree it would make sense for that release.

mattweber commented:
@jpountz Can you please take a look?

jpountz commented:
I'll look now.

jpountz commented:
Query parsing looks good to me.

mikemccand commented:
Thanks @mattweber and @jpountz ... I'll merge this now!

mattweber commented:
Awesome, would you like me to backport to 5x?

mikemccand commented:
> Awesome, would you like me to backport to 5x?

Oh, that would be wonderful, thank you!

mikemccand commented:
@mattweber unfortunately 5.1.0 release is already feature frozen, so this will be in the 5.2.0 release.

mattweber commented:
Ok thanks @mikemccand I will just backport and open new PR against `5.x` branch.  What is the normal process you guys do to backport?  Cherry pick the commit?  For this I might need to go re-fork the `QueryBuilder` from the `5.x` version of lucene.

mattweber commented:
@mikemccand cherry-picked commit applies clean.  I am running tests right now and will open PR for you once those finish up.

mikemccand commented:
@mattweber Yes, cherry pick is normally how I do it; that's great that it's clean!  Thank you.

mikemccand commented:
@mattweber I just pushed the `SynonymGraphFilter` to Lucene (https://issues.apache.org/jira/browse/LUCENE-6664) for its next (6.4.0) feature release; once Lucene releases 6.4.0 and we upgrade to it we should try to un-fork the changes here.

So ... now I'm thinking more about how query parsing could work more generally for any "graph" token streams... and I'm wondering what would actually go wrong if you had used `PositionLengthAttribute > 1` here (instead of `GRAPH_FLAG`) to know when to walk the graph for all paths?

In the cases where old `SynonymFilter` does this, it is in fact a real graph, and it carries important information, i.e. using `GraphQuery` in that case would fix some buggy cases today?

E.g. if I have a synonym:

```
  denial of service -> dos
```

And I try a `match` query with this phrase search:

```
  "denial of service attack"
```

and I'm using the old `SynonymFilter`, it will match `denial of service` and insert `dos` with `posLen=3`, yet today, we will fail to match documents that have `dos attack`, but if we used `GraphQuery` it would match I believe?

mattweber commented:
@mikemccand I believe this will work with any graph token stream.  It does fix some buggy cases of the original `SynonymFilter` which is the reason I added the `GRAPH_FLAG` detection (so it is not a breaking change forcing me to wait until ES6).  [This test](https://github.com/elastic/elasticsearch/blob/master/core/src/test/java/org/elasticsearch/index/search/MatchQueryIT.java#L118) will actually fail if we remove the `GRAPH_FLAG` detection because it will actually traverse the graph from the original `SynonymFilter`.

mikemccand commented:
> @mikemccand I believe this will work with any graph token stream. 

OK this is my impression too (that we do not need `GRAPH_FLAG`, except for back-compat reasons).

Yet, I would argue we are fixing a (nasty, long standing) bug here, by having even the old `SynonymFilter` return correct results?  And that test is indeed showing exactly the bug; thanks for the pointer.  Why not just fix the bug for users still using the old `SynonymFilter` and any other graph token streams?

mattweber commented:
@mikemccand I am all for removing the `GRAPH_FLAG` and just documenting the "fixed" behavior. 

@clintongormley your thoughts?

Also, FYI I started working on support for graph token streams in `query_string` and `simple_query_string` and noticed it introduces a lot of duplicate code because of where we need to override `QueryBuilder` methods.  Are those overrides something we should get into lucene??  I imagine more push back there due to the "fixed" behavior.

mikemccand commented:
>  Are those overrides something we should get into lucene?? I imagine more push back there due to the "fixed" behavior.

That's exactly what I'd like to do!  Patches welcome in Lucene land :)

It's why I was mulling why we can't do this new behavior more generally whenever `posLen > 1` ...

mattweber commented:
@mikemccand Alright, let me work up a patch for lucene.  If we get that in then the only things we need to handle here are the [min should match modifications](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java#L473).

mikemccand commented:
Thanks @mattweber!

mattweber commented:
@mikemccand See [LUCENE-7603](https://issues.apache.org/jira/browse/LUCENE-7603)
